lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/util/AbstractCommandBuilder.java

```diff
@@ -78,26 +81,38 @@ public String buildDriverCommand(final JobSubmissionEvent jobSubmissionEvent) {
     return String.format(this.osCommandFormat, StringUtils.join(commandList, ' '));
   }
 
+  /**
+   * Asembles the command to execute the Evaluator Shim.
+   */
+  public String buildEvaluatorShimCommand(final int evaluatorShimMemory, final String configurationPath) {
+    List<String> commandList = new JavaLaunchCommandBuilder(this.shimLauncherClass, this.commandListPrefix)
+        .setJavaPath(runtimePathProvider.getPath())
+        .setConfigurationFilePaths(Collections.singletonList(configurationPath))
+        .setClassPath(getDriverClasspath())
+        .setMemory(evaluatorShimMemory)
+        .setStandardOut(STD_OUT_FILE)
+        .setStandardErr(STD_ERR_FILE)
+        .build();
+    return String.format(this.osCommandFormat, StringUtils.join(commandList, ' '));
+  }
+
   /**
    * Assembles the command to execute the Evaluator.
    */
   public String buildEvaluatorCommand(final ResourceLaunchEvent resourceLaunchEvent,
                                       final int containerMemory, final double jvmHeapFactor) {
-    List<String> commandList = new ArrayList<>(this.commandListPrefix);
+    List<String> commandList = new ArrayList<>();
     // Use EvaluatorProcess to be compatible with JVMProcess and CLRProcess
     final EvaluatorProcess process = resourceLaunchEvent.getProcess()
-        .setConfigurationFileName(this.reefFileNames.getEvaluatorConfigurationPath())
-        .setStandardErr(this.reefFileNames.getEvaluatorStderrFileName())
-        .setStandardOut(this.reefFileNames.getEvaluatorStdoutFileName());
+        .setConfigurationFileName(this.reefFileNames.getEvaluatorConfigurationPath());
 
     if (process.isOptionSet()) {
       commandList.addAll(process.getCommandLine());
     } else {
-      commandList.addAll(process.setMemory((int) (jvmHeapFactor * containerMemory))
-          .getCommandLine());
+      commandList.addAll(process.setMemory((int) (jvmHeapFactor * containerMemory)).getCommandLine());
     }
 
-    return String.format(this.osCommandFormat, StringUtils.join(getEvaluatorLaunchCommandLine(commandList), ' '));
```

Comment:
I am not sure why this.osCommandForat is removed. I think adding "/bin/sh -c" is no harm.

Also this reminds me of the COMMAND_LIST_PREFIX we used "ln -sf '.' 'reef';", "unzip local.jar;"
It could potentially throw an error if the link already exsits. To update the link, it is "ln -sfn"

Reply:
That's because the evaluator is launched by the shim via the process builder. So there is no need to launch shell and run a command in there.

I'll update the COMMAND_LIST_PREFIX. Good point about the symlink.